javascript/ql/lib/semmle/javascript/frameworks/Spife.qll

```diff
@@ -0,0 +1,416 @@
+  private class HeaderDefinition extends Http::ExplicitHeaderDefinition, DataFlow::MethodCallNode instanceof ReplySource {
+    HeaderDefinition() {
+      // reply.header(RESPONSE, 'Cache-Control', 'no-cache')
+      exists(DataFlow::MethodCallNode call |
+        this.ref() = call and
+        call.getMethodName() = "header" and
+        call.getNumArgument() = 3
+      )
+    }
+
+    override predicate definesHeaderValue(string headerName, DataFlow::Node headerValue) {
+      // reply.header(RESPONSE, 'Cache-Control', 'no-cache')
+      headerName = this.getNameNode().getStringValue() and
+      headerValue = this.getArgument(2)
+    }
+
+    override DataFlow::Node getNameNode() { result = this.getArgument(1) }
+
+    override RouteHandler getRouteHandler() { result = this.getRouteHandler() }
+  }
+
+  /**
+   * An invocation that sets any number of headers of the HTTP response.
+   */
+  private class MultipleHeaderDefinitions extends Http::ExplicitHeaderDefinition, DataFlow::CallNode {
+    ReplySource reply;
+
+    MultipleHeaderDefinitions() {
+      // reply.header(RESPONSE, {'Cache-Control': 'no-cache'})
+      // reply(RESPONSE, {'Cache-Control': 'no-cache'})
+      reply.ref().(DataFlow::CallNode).getCalleeName() = ["header", "reply"] and
+      reply.ref().(DataFlow::CallNode).getAnArgument().getALocalSource() instanceof
+        DataFlow::ObjectLiteralNode and
+      this = reply
+    }
+
+    /**
+     * Gets a reference to the multiple headers object that is to be set.
+     */
+    private DataFlow::ObjectLiteralNode getAHeaderSource() {
+      result = this.getAnArgument().getALocalSource()
+    }
+
+    override predicate definesHeaderValue(string headerName, DataFlow::Node headerValue) {
+      exists(string header |
+        this.getAHeaderSource().hasPropertyWrite(header, headerValue) and
+        headerName = header.toLowerCase()
+      )
+    }
+
+    override DataFlow::Node getNameNode() {
+      result = this.getAHeaderSource().getAPropertyWrite().getPropertyNameExpr().flow()
+    }
+
+    override RouteHandler getRouteHandler() { result = reply.getRouteHandler() }
+  }
+
+  /**
+   * A header produced by a route handler with no explicit declaration of a Content-Type.
+   */
+  private class ContentTypeRouteHandlerHeader extends Http::ImplicitHeaderDefinition,
+    DataFlow::FunctionNode instanceof RouteHandler {
+    override predicate defines(string headerName, string headerValue) {
+      headerName = "content-type" and headerValue = "application/json"
+    }
+
+    override Http::RouteHandler getRouteHandler() { result = this }
+  }
+
+  /**
+   * An HTTP cookie defined in a Spife HTTP response.
+   */
+  private class CookieDefinition extends Http::CookieDefinition, DataFlow::MethodCallNode instanceof ReplySource {
+    CookieDefinition() {
+      // reply.cookie(RESPONSE, 'TEST', 'FOO', {"maxAge": 1000, "httpOnly": true, "secure": true})
+      this.ref().(DataFlow::MethodCallNode).getMethodName() = "cookie"
```

Comment:
That won't work.  

You want a method called on a reply, not a reply that is a method-call. 
Also, the reply isn't a definition of a cookie, the call to `cookie` is.  

```suggestion
  private class CookieDefinition extends Http::CookieDefinition, DataFlow::MethodCallNode {
    CookieDefinition() {
      // reply.cookie(RESPONSE, 'TEST', 'FOO', {"maxAge": 1000, "httpOnly": true, "secure": true})
      this = any(ReplySource r).ref().getAMethodCall("cookie")
```